tidy3d/components/geometry/polyslab.py

```diff
@@ -1406,190 +1406,214 @@ def _surface_area(self, bounds: Bound) -> float:
+        integral is evaluated on every retained patch; the resulting derivative
+        is split equally between the two vertices that bound the edge segment.
+        """
+        # rmin/rmax over the geometry ∩ simulation box
+        rmin, rmax = derivative_info.bounds_intersect
 
-        rmin, rmax = derivative_info.bounds
         ax_min, (r1_min, r2_min) = self.pop_axis(rmin, axis=self.axis)
         ax_max, (r1_max, r2_max) = self.pop_axis(rmax, axis=self.axis)
 
-        num_1 = 1 if r1_min == r1_max else _NUM_PTS_DIM_SLAB_BOUNDS
-        num_2 = 1 if r2_min == r2_max else _NUM_PTS_DIM_SLAB_BOUNDS
-
-        num_cells = num_1 * num_2
-        ones = np.ones(num_cells)
-        zeros = np.zeros(num_cells)
-
-        def meshgrid_flatten_stack(*args) -> np.ndarray:
-            """Take set of `d` coords, meshgrid them, flatten, and assemble in `(N, d)` array."""
-            coords = np.meshgrid(*args, indexing="ij")
-            coords = [c.flatten() for c in coords]
-            return np.stack(coords, axis=-1)
-
-        # get center points and areas
-        r1_centers = np.linspace(r1_min, r1_max, 2 * num_1 + 1)[1::2]
-        r2_centers = np.linspace(r2_min, r2_max, 2 * num_2 + 1)[1::2]
-        planar_centers = meshgrid_flatten_stack(r1_centers, r2_centers)
-
-        area = 1.0
-        for rmin, rmax in zip((r1_min, r2_min), (r1_max, r2_max)):
-            if rmin != rmax:
-                area *= rmax - rmin
-
-        areas = area * np.ones(num_cells) / num_cells
-
-        def get_grad(min_max_index: int) -> float:
-            """Compute gradient for either min or max dimension."""
-
-            # select the normal sign and the axis value
-            if min_max_index == 0:
-                ax_val = ax_min
-            else:
-                ax_val = ax_max
-
-            edge_centers_xyz = self.unpop_axis_vect(ones * ax_val, planar_centers)
-
-            # construct basis functions
-            normals = self.unpop_axis_vect(ones, np.stack((zeros, zeros), axis=-1))
-            perps1 = self.unpop_axis_vect(zeros, np.stack((ones, zeros), axis=-1))
-            perps2 = self.unpop_axis_vect(zeros, np.stack((zeros, ones), axis=-1))
-
-            rr1, rr2, axx_val = meshgrid_flatten_stack(r1_centers, r2_centers, ax_val).T
-
-            xx, yy, zz = self.unpop_axis_vect(axx_val, np.stack((rr1, rr2), axis=-1)).T
-
-            inside = self.inside(xx, yy, zz).squeeze().flatten()
-
-            areas_masked = areas * inside
+        # planar grid resolution, clipped to polygon bounding box
+        extents = np.abs([r1_max - r1_min, r2_max - r2_min])
+        if extents[0] == 0 or extents[1] == 0:
+            return 0.0
+        n1, n2 = np.maximum(1, np.ceil(extents / _VJP_SAMPLE_SPACING).astype(int))
+
+        # build shapely polygon & its bbox
+        face_verts = self.base_polygon if min_max_index == 0 else self.top_polygon
+        face_poly = shapely.Polygon(face_verts).buffer(fp_eps)
```

Comment:
run into issues if both the `verices` and `slab_bounds` are traced with shapely not liking tracers?